### PR TITLE
Add CHANGELOG.md

### PR DESCRIPTION
**Pull request by hellofromtonya**

### CODE REVIEW

**Comment by hellofromtonya** [comment]
@iCaspar @christophherr Go through the changelog.  See if I captured everything.

**Comment by hellofromtonya** [comment]
Excellent.  I’ll add more to it in order to:

1. Emphasis significance of the auditing and compliance work we all did.
2. Listing of the new functions.
3. List of deprecated functions.
4. Plus, list previous releases from the release logs.

Get it done. And then, we are releasing. WooHoo!

**Comment by iCaspar** [comment]
Deprecated functions: `api/widgets/beans_widget_area()` -> `beans_get_widget_area_output()`
Fixed: `beans_get_widget_area()` and `beans_get_widget()`

File: CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+* Fixed Actions API to allow priority of 0 to be modified
+* Fixed Actions API double subhook calls
+* Fixed `beans_path_to_url` to bail out when relative URL
+* Fixed count for `beans_count_recursive`
+* Fixed removing tilde from `beans_url_to_path`
+* Fixed processing relative URLs in `beans_url_to_path`
+* Fixed altering of non-internal URLs in `beans_url_to_path`
+* Fixed `beans_get_layout_class` not returning correct classes when secondary is no longer registered
+* Fixed 'Next Post' icon close markup ID
+* Fixed 'Read More' icon markup IDs
```
Comment (iCaspar) [inline CHANGELOG.md]: See my comment on the PR for 2 api/widget functions that were fixed, to add here. The fix for both was "not always returning false on fail".

**Comment by hellofromtonya** [comment]
@christophherr and @iCaspar I've insignificantly improved the changelog:

1. Broke it out into different categories.
2. Walked through the codebase and captured the information for the relative categories.
3. Rolled in your comments above.
4. Added past releases with links to their changelogs.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,32 +1,166 @@
...
-* Fixed 'Read More' icon markup IDs
-* Eliminated storing actions in encoded strings
+## 2018.06.21 - version 1.5.0-beta
+
+This pre-release includes security, web accessibility, code quality improvements, performance improvements, and bug fixes.  The entire framework is now WPCS compliant.  The APIs are now fully and thoroughly well-tested. All found issues are resolved.  New functionality is added. 
+
+### Improved
+* Improved web accessibility by adding skip links, ARIA, labels, tabs, and more.
+* Improved the performance of the APIs by reducing expensive processes, memory usage, and execution time.
+* Improved the APIs by adding a thorough unit and integration test suites.
```
Comment (christophherr) [inline CHANGELOG.md]: adding thorough

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,32 +1,166 @@
...
+* Added `beans_scandir()` to scan the given directory path, lists all the files and directories, and removes the `.` and `..` files.
+* Added `beans_uikit_get_all_components()` to check all of the UIkit components for the given type.
+* Added `beans_uikit_get_all_dependencies()` to check all of the UIkit dependencies for the given component(s).
+* Added `beans_get_widget_area_output()` to replace `beans_get_widget_area()`.
+* Added `beans_add_compiler_options_to_settings()` to instantiate `_Beans_Compiler_Options` through a hook instead on file load.
+* Added `beans_add_page_assets_compiler()` to instantiate `__Beans_Page_Compiler` through a hook instead on file load.
+* Added `beans_add_image_options_to_settings()` to instantiate `___Beans_Image_Options` through a hook instead on file load.
+* Added `beans_has_primary_sidebar()` to check if the given layout has a primary sidebar.
+* Added `beans_has_secondary_sidebar()` to check if the given layout has a secondary sidebar.
+* Added many new private functions.
```
Comment (christophherr) [inline CHANGELOG.md]: Added `beans_skip_links_list` to filter the skip links.

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,32 +1,166 @@
...
+
+### Changes
+* `beans_add_action()` returns `false` when the action is not added via `add_action`.
+* `beans_add_smart_action()` returns `false` when the action is not added via `add_action`.
+* `beans_modify_action()` returns `false` when no hook, callback, priority, and args are not given.
```
Comment (christophherr) [inline CHANGELOG.md]: are given. (or drop the `no` before hook)

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,32 +1,166 @@
...
+* `beans_modify_action_hook()` returns `false` when the hook is empty or not a string.
+* `beans_modify_action_callback()` returns `false` when the callback is empty.
+* `beans_replace_action()` returns `false` when no hook, callback, priority, and args are not given.
```
Comment (christophherr) [inline CHANGELOG.md]: See previous comment.

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,32 +1,166 @@
...
+* `beans_replace_action_hook()` returns `false` when the hook is empty or not a string.
+* `beans_replace_action_callback()` returns `false` when the callback is empty.
+* `beans_remove_action()` sets "removed" to the default when no current action exists.
+* `beans_reset_action()` bails out if the action does not need to be reset.
+* Unset the global variable in `beans_selfclose_markup()` to reduce memory usage.
+* `beans_wrap_markup()` bails out and return false if the given `$tag` is empty.
```
Comment (christophherr) [inline CHANGELOG.md]: returns

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,32 +1,166 @@
...
+* `beans_wrap_inner_markup()` bails out and return false if the given `$tag` is empty.
```
Comment (christophherr) [inline CHANGELOG.md]: returns

File: CHANGELOG.md [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+* Fixed `beans_get_layout_class()` not returning correct classes when secondary is no longer registered.
+* Fixed 'Next Post' icon close markup ID.
+* Fixed 'Read More' icon markup IDs.
+* Fixed `beans_get_widget_area()` to return `false` on fail.
+* Fixed `beans_get_widget()` to return `false` on fail.
+* Fixes to private functions and methods are not noted here.
+
+### Deprecated
+* Deprecated `beans_count_recursive()` as it is unused in Beans.
+* Deprecated `beans_widget_area()` by replacing it with a renamed function `beans_get_widget_area_output()` that better describes the expected behavior.
```
Comment (christophherr) [inline CHANGELOG.md]: We also deprecated `The image toolbar's dashicons' class attributes`

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,32 +1,166 @@
...
-* Eliminated storing actions in encoded strings
+## 2018.06.21 - version 1.5.0-beta
+
+This pre-release includes security, web accessibility, code quality improvements, performance improvements, and bug fixes.  The entire framework is now WPCS compliant.  The APIs are now fully and thoroughly well-tested. All found issues are resolved.  New functionality is added. 
+
+### Improved
+* Improved web accessibility by adding skip links, ARIA, labels, tabs, and more.
+* Improved the performance of the APIs by reducing expensive processes, memory usage, and execution time.
+* Improved the APIs by adding a thorough unit and integration test suites.
+* Improved the APIs by adding type hinting, reducing redundant code, and removing used code.
```
Comment (iCaspar) [inline CHANGELOG.md]: ... removing unused code.

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,32 +1,166 @@
...
+* Improved Beans by making it WPCS compliant.
+* Improved CSS compiler to have one block of declarations per line.
+* Improved security by auditing and applying escaping and sanitizing best practices.
+* Improved the Actions API by eliminating storing actions in encoded strings.
```
Comment (iCaspar) [inline CHANGELOG.md]: ...by no longer storing...

File: CHANGELOG.md [hunk 11]
```diff
@@ -1,32 +1,166 @@
...
+### Updated
+* Updated the language's tm-beans.pot file.
+* Updated UIkit to version 2.27.5 (see [UIkit changelog](https://github.com/uikit/uikit/blob/v2/master/CHANGELOG.md#2275) for more information).
+
+### Added
+* Added `beans_str_ends_with()` to check if the given string ends with the given substring(s).
+* Added `beans_str_starts_with()` to check if the given string starts with the given substring(s).
+* Added `beans_multi_array_key_exists()` to check if a key or index exists in a multi-dimensional array.
+* Added `beans_join_arrays()` to join two arrays together.
+* Added `beans_array_unique()` to remove duplicate values and re-indexes the array.
```
Comment (iCaspar) [inline CHANGELOG.md]: ... and re-index ...

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,32 +1,166 @@
...
+* Added `beans_join_arrays_clean()` to join two arrays, removes the duplicate values and empties, and provides an option to re-index the clean array.
```
Comment (iCaspar) [inline CHANGELOG.md]: ...arrays, remove the .... and provide an ...

File: CHANGELOG.md [hunk 13]
```diff
@@ -1,32 +1,166 @@
...
+* Added `beans_scandir()` to scan the given directory path, lists all the files and directories, and removes the `.` and `..` files.
```
Comment (iCaspar) [inline CHANGELOG.md]: ... path, list all ... and remove the ...

File: CHANGELOG.md [hunk 14]
```diff
@@ -1,32 +1,166 @@
...
+* Added `beans_uikit_get_all_components()` to check all of the UIkit components for the given type.
+* Added `beans_uikit_get_all_dependencies()` to check all of the UIkit dependencies for the given component(s).
+* Added `beans_get_widget_area_output()` to replace `beans_get_widget_area()`.
+* Added `beans_add_compiler_options_to_settings()` to instantiate `_Beans_Compiler_Options` through a hook instead on file load.
```
Comment (iCaspar) [inline CHANGELOG.md]: ... instead of on ...

File: CHANGELOG.md [hunk 15]
```diff
@@ -1,32 +1,166 @@
...
+* Added `beans_add_page_assets_compiler()` to instantiate `__Beans_Page_Compiler` through a hook instead on file load.
```
Comment (iCaspar) [inline CHANGELOG.md]: ... instead of on ...

File: CHANGELOG.md [hunk 16]
```diff
@@ -1,32 +1,166 @@
...
+* Added `beans_add_image_options_to_settings()` to instantiate `___Beans_Image_Options` through a hook instead on file load.
```
Comment (iCaspar) [inline CHANGELOG.md]: ... instead of on ...

**Comment by iCaspar** [comment]
This version looks much better! It reflects more accurately the massive amount of work that's been done!

File: CHANGELOG.md (hunk 9 above)
Comment (hellofromtonya) [reply CHANGELOG.md]: LOL Boy, I blew that one!

Comment (christophherr) [reply CHANGELOG.md]: We are all about lean and mean...

Comment (hellofromtonya) [reply CHANGELOG.md]: We only leave in the unused code. LOL

**Comment by hellofromtonya** [comment]
Okay, @iCaspar and @christophherr all of your notes are now in the latest.  Are we ready?  If yes, go ahead and approve.  If no, keep the changes flowing.

**Comment by christophherr** [comment]
Looks good to me.

**Pull request merged**